Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+{
+    VERIFY(!insn.modrm().is_register());
+    set_c1(0);
+    i64 value = static_cast<i64>(fpu_pop());
+    // FIXME: Respect shadow values
+    insn.modrm().write64(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u64>(value)));
+}
+
+void SoftFPU::FBLD_M80(const X86::Instruction&) { TODO_INSN(); }
+void SoftFPU::FBSTP_M80(const X86::Instruction&) { TODO_INSN(); }
+
+void SoftFPU::FXCH(const X86::Instruction& insn)
+{
+    // FIXME: implicit argument `D9 C9` -> st[0] <-> st[1]?
+    VERIFY(insn.modrm().is_register());
+    set_c1(0);
+    auto tmp = fpu_get(0);
+    fpu_set(0, fpu_get(insn.modrm().register_index()));
+    fpu_set(insn.modrm().register_index(), tmp);
+}
+
+void SoftFPU::FCMOVE(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    if (m_cpu.zf())
+        fpu_set(0, fpu_get(insn.modrm().rm()));
+}
+void SoftFPU::FCMOVNE(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    if (!m_cpu.zf())
+        fpu_set(0, fpu_get((insn.modrm().reg_fpu())));
+}
+
+void SoftFPU::FCMOVB(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    if (m_cpu.cf())
+        fpu_set(0, fpu_get(insn.modrm().rm()));
+}
+void SoftFPU::FCMOVNB(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m32int = (i32)insn.modrm().read32(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    fpu_push((long double)m32int);
+}
+void SoftFPU::FCMOVBE(const X86::Instruction& insn)
+{
+    if (m_cpu.cf() | m_cpu.zf())
+        fpu_set(0, fpu_get(insn.modrm().rm()));
+}
+void SoftFPU::FCMOVNBE(const X86::Instruction& insn)
+{
+    if (!(m_cpu.cf() | m_cpu.zf()))
+        fpu_set(0, fpu_get(insn.modrm().rm()));
+}
+
+void SoftFPU::FCMOVU(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    if (m_cpu.pf())
+        fpu_set(0, fpu_get((insn.modrm().reg_fpu())));
+}
+void SoftFPU::FCMOVNU(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    if (m_cpu.pf())
+        fpu_set(0, fpu_get((insn.modrm().reg_fpu())));
+}
+
+// BASIC ARITHMETIC
+void SoftFPU::FADD_RM32(const X86::Instruction& insn)
+{
+    // XXX look at ::INC_foo for how mem/reg stuff is handled, and use that here too to make sure this is only called for mem32 ops
+    if (insn.modrm().is_register()) {
+        fpu_set(0, fpu_get(insn.modrm().register_index()) + fpu_get(0));
+    } else {
+        auto new_f32 = insn.modrm().read32(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f32 = bit_cast<float>(new_f32.value());
+        fpu_set(0, fpu_get(0) + f32);
+    }
+}
+void SoftFPU::FADD_RM64(const X86::Instruction& insn)
+{
+    // XXX look at ::INC_foo for how mem/reg stuff is handled, and use that here too to make sure this is only called for mem64 ops
+    if (insn.modrm().is_register()) {
+        fpu_set(insn.modrm().register_index(), fpu_get(insn.modrm().register_index()) + fpu_get(0));
```

Comment:
FADD shouldn't be functionally different for having a larger argument size, right? Why does RM32 use index 0 as target when this has selectable target?

Reply:
everything seems alright, intel is just weird